src/qfusion.cpp

```diff
@@ -485,17 +419,50 @@ bitCapInt QFusion::ForceMReg(bitLenInt start, bitLenInt length, bitCapInt result
     return qReg->ForceMReg(start, length, result, doForce);
 }
 
+void QFusion::BufferArithmetic(
+    bitLenInt* controls, bitLenInt controlLen, int toAdd, bitLenInt inOutStart, bitLenInt length)
+{
+    // We can fuse arithmetic, but this does not necessarily commute with nonarithmetic gates.
+    // We must flush the bit buffers, if they aren't arithmetic buffers.
+
+    bitLenInt i;
+
+    for (i = 0; i < controlLen; i++) {
+        FlushBit(controls[i]);
+    }
+
+    BitBufferPtr toCheck;
+    BitBufferPtr bfr = std::make_shared<ArithmeticBuffer>(false, controls, controlLen, inOutStart, length, toAdd);
+
+    for (i = 0; i < length; i++) {
+        toCheck = bitBuffers[inOutStart + i];
+        if (!(bfr->CompareControls(toCheck))) {
+            FlushReg(inOutStart, length);
+            break;
+        }
+    }
+
+    toCheck = bitBuffers[inOutStart];
+    if (toCheck == NULL) {
+        for (i = 0; i < length; i++) {
+            bitBuffers[inOutStart + i] = bfr;
+        }
+        for (i = 0; i < controlLen; i++) {
+            bitControls[controls[i]].push_back(inOutStart);
+        }
+    } else {
+        dynamic_cast<ArithmeticBuffer*>(toCheck.get())->toAdd += toAdd;
```

Comment:
This cast makes me a bit nervous.  Can you add some more comments in this code in general, especially what each of the blocks is doing, but also specifically here as to why this is safe to do?

Reply:
"CompareControls" started as a method to compare control bits, to determine whether buffers could be combined. However, it also took over checking whether buffer "isArithmetic." I've changed the method name to "Combinable" to make this clearer, and I added a comment, here.

Once a buffer passes "Combinable," here, we know it's either an arithmetic buffer that can be combined, or NULL. Below, we check that the buffer is not NULL, in which case it's definitely a "combinable" arithmetic buffer.

Reply:
Instead of deep fiddling, add an accessor method and change the cast to get rid of the get() call.